crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,326 @@
+//!
+//! As we have seen above, visibility constraints can apply outside of a control flow element.
+//! So we need to consider the possibility that multiple constraints apply to the same binding.
+//! Here, we consider what happens if multiple `if`-statements lead to a sequence of constraints.
+//! Consider the following example:
+//! ```py
+//! x = 0
+//!
+//! if test1:
+//!     x = 1
+//!
+//! if test2:
+//!     x = 2
+//! ```
+//! The binding `x = 2` is easy to analyze. Its visibility corresponds to the truthiness of `test2`.
+//! For the `x = 1` binding, things are a bit more interesting. It is always visible if `test1` is
+//! always-true *and* `test2` is always-false. It is never visible if `test1` is always-false *or*
+//! `test2` is always-true. And it is ambiguous otherwise. This corresponds to a ternary *test1 AND
+//! ~test2* operation in three-valued Kleene logic [Kleene]:
+//!
+//!       | AND          | always-false | ambiguous    | always-true  |
+//!       |--------------|--------------|--------------|--------------|
+//!       | always false | always-false | always-false | always-false |
+//!       | ambiguous    | always-false | ambiguous    | ambiguous    |
+//!       | always true  | always-false | ambiguous    | always-true  |
+//!
+//! The `x = 0` binding can be handled similarly, with the difference that both `test1` and `test2`
+//! are negated:
+//! ```py
+//! x = 0  # ~test1 AND ~test2
+//!
+//! if test1:
+//!     x = 1  # test1 AND ~test2
+//!
+//! if test2:
+//!     x = 2  # test2
+//! ```
+//!
+//! ### Merged constraints (ternary OR)
+//!
+//! Finally, we consider what happens in "parallel" control flow. Consider the following example
+//! where we have omitted the test condition for the outer `if` for clarity:
+//! ```py
+//! x = 0
+//!
+//! if <…>:
+//!     if test1:
+//!         x = 1
+//! else:
+//!     if test2:
+//!         x = 2
+//!
+//! use(x)
+//! ```
+//! At the usage of `x`, i.e. after control flow has been merged again, the visibility of the `x =
+//! 0` binding behaves as follows: the binding is always visible if `test1` is always-false *or*
+//! `test2` is always-false; and it is never visible if `test1` is always-true *and* `test2` is
+//! always-true. This corresponds to a ternary *OR* operation in Kleene logic:
+//!
+//!       | OR           | always-false | ambiguous    | always-true  |
+//!       |--------------|--------------|--------------|--------------|
+//!       | always false | always-false | ambiguous    | always-true  |
+//!       | ambiguous    | ambiguous    | ambiguous    | always-true  |
+//!       | always true  | always-true  | always-true  | always-true  |
+//!
+//! Using this, we can annotate the visibility constraints for the example above:
+//! ```py
+//! x = 0  # ~test1 OR ~test2
+//!
+//! if <…>:
+//!     if test1:
+//!         x = 1  # test1
+//! else:
+//!     if test2:
+//!         x = 2  # test2
+//!
+//! use(x)
+//! ```
+//!
+//! ### Explicit ambiguity
```

Comment:
I need to look at this section again. I'm not sure if that makes sense. Or if there is another way to solve this without explicit ambiguity constraints.

Reply:
I think it makes sense? On a statically-analyzable branch, we have to add `test` visibility to the branch path and `~test` visibility to the not-branch path. `VisibilityConstraint::Ambiguous` is just a shortcut to do the same thing in the case where there is no `test` expression we plan to analyze for truthiness later. It's "the same thing" because `ambiguous` and `~ambiguous` are indistinguishable; we can just add `VisibilityConstraint::Ambiguous` to both paths.